Document all changes to MongoDB CDC Event Flattening (#228)
debezium/debezium.github.io: docs/configuration/mongodb-event-flattening.asciidoc

This option allows you to convert a hierarchical document into a flat structure suitable for a table-like storage.

=== MongoDB `$unset` handling

> MongoDB allows you to make `$unset` operations which allows you to remove a certain field from a Document, and since the collections are schemaless it becomes hard to find a way to tell the consumers/sinkers which a field is now missing, the approach debezium uses is to set the desired to remove field to null value.

Reviewer: s/debezium/Debezium/
```suggestion
MongoDB allows you to run `$unset` operations which allows you to remove a certain field from a document, and since the collections are schemaless, it becomes hard to find a way to tell the consumers/sinkers which a field is now missing, the approach Debezium uses is to set the desired to remove field to null value.
```

Given the operation
[source,json,indent=0]
----
{
    "after":null,
    "patch":"{\"$unset\" : {\"a\" : true}}"
}
----

The final encoding will look like
[source,json,indent=0]
----
{
    "id": 1,
    "a": null
}
----

Note that other mongo operations might cause an `$unset` internally, `$rename` is one example.

=== Determine original operation

> When a message is flattened the final result won't show weather it was an insert, update or first read (Deletions can be detected via tombstones or rewrites, see link:#configuration_options[Configuration options]).

Reviewer: ```suggestion
When a message is flattened the final result won't show whether it was an insert, update or first read (deletions can be detected via tombstones or rewrites, see link:#configuration_options[Configuration options]).
```

> To solve this problem Debezium offers an option to propagate the original operation as an added header to the message.

Reviewer: ```suggestion
To solve this problem Debezium offers an option to propagate the original operation via a header added to the message.
```

To enable this feature the option `operation.header` must be set to `true`.

[source]
----
transforms=unwrap,...
transforms.unwrap.type=io.debezium.connector.mongodb.transforms.UnwrapFromMongoDbEnvelope
transforms.unwrap.operation.header=true
----

> Possible values can be as in the link:docs/connectors/mongodb/#change-events-value[MongoDB connector change events]

Reviewer: ```suggestion
The possible values are the ones from the `op` field of link:docs/connectors/mongodb/#change-events-value[MongoDB connector change events].
```

== Configuration options
[cols="35%a,10%a,55%a",width=100,options="header,footer",role="table table-bordered table-striped"]
|=======================
|Property
|Default
|Description

> |`array.encoding`

Reviewer: Very nice, thanks for covering the existing ones, too. 👍

|`array`
|The SMT converts MongoDB arrays into arrays as defined by Apache Connect (or Apache Avro) schema.

|`flatten.struct`
|`false`
|The SMT flattens structs by concatenating the fields into plain properties, using a configurable delimiter.

|`flatten.struct.delimiter`
|`_`

> |Delimiter to concat between field names from the input record when generating field names for the output record. Only works when `flatten.struct` is `true`

Reviewer: ```suggestion
|Delimiter to concat between field names from the input record when generating field names for the output record. Only applies when `flatten.struct` is set to `true`.
```

|`operation.header`
|`false`
|The SMT adds the link:docs/connectors/mongodb/#change-events-value[event operation] as a message header.

|`drop.tombstones`
|`true`
|The SMT removes the tombstone generated by Debezium from the stream.

|`delete.handling.mode`
|`drop`

> |The SMT can `drop`, `rewrite` or pass record (`none`). Rewrite mode will add `__deleted` column with true/false values based on record operation.

Reviewer: ```suggestion
|The SMT can `drop`, `rewrite` or pass delete records (`none`). The `rewrite` mode will add a `__deleted` field set to `true` or `false` depending on the represented operation.
```

> |The SMT can `drop`, `rewrite` or pass record (`none`). Rewrite mode will add `__deleted` column with true/false values based on record operation.

Reviewer: ```suggestion
|The SMT can `drop`, `rewrite` or pass delete records (`none`). The `rewrite` mode will add a `__deleted` field set to `true` or `false` depending on the represented operation.
```
